### ISSUE TITLE
feat: allow using abstract class as identifier in Container.set

### ISSUE THREAD

**Issue by domi7777** [typestack/typedi#144]
**I was trying to...**
...create an abstraction layer for our project and noticed some discrepancy, example:

```typescript
import {Container} from 'typedi';

abstract class Animal {
    abstract getSound(): string;
}

class Cat extends Animal {
    constructor(private sound: string) {
        super();
    }

    getSound(): string {
        return this.sound;
    }
}

Container.set(Animal, new Cat('meow')); // compiles

console.log(Container.get(Animal).getSound()); // does not compile
```
I wonder why `Container.set()` with abstract class "token" does compile but Container.get() not, because it expects a class that can be instantiated (ObjectType)

And:

```typescript
console.log(Container.get(Animal as unknown as ObjectType<Animal>).getSound());
```

Compiles and run fine.

**The problem:**
- Is there any reason why we can't use abstract class as service token?
- Am I misusing `typedi`?

Thanks for your help & efforts for this great lib.

**Comment by NoNameProvided** [comment]
Hi!

You are not misusing the library, this behavior is a result of a not correctly set type on the parameters. What is your expected behavior? Would you like to provide anything as a key?

**Comment by domi7777** [comment]
Thanks for answering, my question is, considering the code above, why does this code compile:

```Container.set(Animal, new Cat('meow'));```

but this one doesn't:

```Container.get(Animal);```

Here's the TS error I'm getting:

![image](https://user-images.githubusercontent.com/5888729/104027870-4d78bf00-51c8-11eb-8f87-b133986629c3.png)

**And** if I just bypass the compilation for this line, it works fine at runtime.

So I guess it has to do with the fact that `Animal` is an abstract class. However, I'd like to use as a key, so that I don't have to be aware of the actual implementation details when I'm using it. Does that make sense?

**Comment by NoNameProvided** [comment]
Fixed in 60b3dc96858fd6867ca4a9c75ead729c54befc22. In the next version you will be able to set abstract classes as keys.

**Comment by domi7777** [comment]
Thank you, I'll try it. Closing this ticket.

**Comment by NoNameProvided** [comment]
Reopening until fix released.

**Comment by NoNameProvided** [comment]
This has been released as `0.9.0`.
